### ISSUE TITLE
Path-guarded setups: a command-shape guard that inspects the full command string blocks `cd`-then-run even when the target directory is permitted, and can also block a blocked command merely quoted in a log entry

### ISSUE THREAD

**Issue by Pfeifdanix** [rebelytics/one-skill-to-rule-them-all#171]
## Environment

- macOS, bash 3.2.57 (`/bin/bash`)
- Claude Code, with a PreToolUse hook that inspects the literal Bash
  command text for governed path segments combined with a write
  indicator (`=`, `>`, `cp`, `mv`, etc.) or specific denylisted words, and
  denies the whole call when both are present anywhere in the string —
  independent of whether the path is actually being written to.

## Symptom

`references/environments.md`, "If CLAUDE.md (or the equivalent config) is
governance-protected", documents what to do when an EDIT to the
governance-protected config file is denied. It does not cover a related
but distinct case: running one of the skill's OWN bundle scripts (the
validator, the migration script) while the current working directory, or
a `cd` in the same command, touches the governed path — even when the
script call itself is read-only and the destination directory is one the
guard is meant to allow reads from.

Two concrete failures observed while staging an update to this skill
under such a guard:

1. `cd ~/.claude/skills/task-observer && python3 scripts/validate-skill-bundle.py .`
   was refused outright. The guard's rule matched on the combination of
   the governed path segment (`~/.claude/`) and the change-of-directory
   plus relative-path shape of the command, not on any actual write —
   the validator only reads and, with `--pack`, writes a `.skill` file
   in the CURRENT directory, which here would have been the guarded one
   only because of the `cd`.
2. Separately, writing an observation file whose Issue/Suggested-improvement
   text quoted the blocked command verbatim (to document what had failed)
   was ALSO denied — the same guard evidently inspects the full text of
   any Bash-tool call, including one whose payload is a `Write` to an
   observation file that merely contains the blocked string as a quoted
   example, not as something being executed.

Neither of these is "the config file itself is protected", which is what
the existing section addresses. Both are more general: (1) a script
invocation shape that happens to resemble a governed write because of a
directory change, and (2) a textual guard that cannot distinguish
"execute this" from "quote this for the record".

## Reproduction

1. Configure a PreToolUse-style hook (or any command-shape guard) that
   denies a Bash call whenever the command string contains both a
   governed path segment and a change-of-directory token, regardless of
   what runs after the `cd`.
2. Attempt to validate a staged skill bundle by changing into the skill
   directory first: `cd /path/under/governed/tree && python3 scripts/validate-skill-bundle.py .`
   — denied, even though the script only reads files and the guard's
   stated purpose is to prevent writes to the governed tree.
3. Separately, attempt to log an observation (a `Write` call) whose body
   text contains the exact blocked command string from step 2, quoted as
   an example of what failed — also denied by a guard that scans the
   whole command payload for the governed path.

## Proposed change

In `references/environments.md`, immediately after "If CLAUDE.md (or the
equivalent config) is governance-protected" and its "blocked hook
installation" corollary, add a third case:

```markdown
**A third case: invoking the skill's own bundle scripts under a
command-shape guard.** Some guards inspect the full command STRING rather
than only the target file, and deny any call where a governed path segment
appears together with a directory-change token — independent of whether
the command that follows is a write. A `cd` into the skill directory
before running the validator or migration script triggers this even when
the script itself only reads. The fix is mechanical, not a fallback to
ask the user: call bundle scripts by ABSOLUTE PATH from wherever the agent
already is, and never `cd` into a guarded directory first —
`python3 /absolute/path/to/scripts/validate-skill-bundle.py /absolute/path/to/skill-dir`
keeps the governed segment visible as a plain argument instead of hiding
it behind a working-directory change, which is what such guards actually
key on. This applies to every bundled script (`validate-skill-bundle.py`,
`migrate-log.py`) in every environment where the skills directory sits
under a path a hook or file-protection rule inspects by string content.

**The same guard often inspects text, not only executed commands.** A
command-shape guard that scans the full text of a tool call can also
reject a blocked command quoted verbatim inside an observation's Issue or
Suggested-improvement field — the write is to an unrelated file (the
observation log), but the guard cannot tell "about to execute this" from
"quoting this as a record of what was rejected". When logging an
observation about a blocked command, describe what was attempted and why
it was rejected in prose, rather than reproducing the literal command
text, so the log entry itself is not re-blocked by the same guard it is
documenting.
```

## Why this matters generally

Any setup with a PreToolUse-style hook or file-protection rule that
classifies commands by their full text — not only by the file ultimately
touched — will hit both failure shapes: a legitimate read-only script call
shaped like a write (via `cd`), and a legitimate log entry shaped like the
blocked command it describes. Neither is specific to one hook
implementation; both follow from "the guard reads the command as a
string", which is a common and reasonable way to build such a guard.

**Comment by rebelytics** [comment]
Shipped in v3.3.0 — the guard covers running a bundled script from a permitted directory, not just editing a protected file. Thanks!
